### PR TITLE
Expose more of `WorkingBeatmap` via interface

### PR DESCRIPTION
**Pull request by peppy**
This exposes all methods and metadata required to propagate `IWorkingBeatmap` usage out further (around 90% of usages game-wide).

### CODE REVIEW

File: osu.Game/Beatmaps/WorkingBeatmap.cs [hunk 1]
```diff
@@ -28,8 +28,10 @@ public abstract class WorkingBeatmap : IWorkingBeatmap
     {
         public readonly BeatmapInfo BeatmapInfo;
 
+        // ReSharper disable once FieldHidesInterfacePropertyWithDefaultImplementation
         public readonly BeatmapSetInfo BeatmapSetInfo;
 
+        // ReSharper disable once FieldHidesInterfacePropertyWithDefaultImplementation
         public readonly BeatmapMetadata Metadata;
```
Comment (peppy) [inline osu.Game/Beatmaps/WorkingBeatmap.cs:35]: Not sure about this. We can remove the default implementation if preferred.

File: osu.Game/Beatmaps/IWorkingBeatmap.cs [hunk 2]
```diff
@@ -17,33 +17,69 @@ namespace osu.Game.Beatmaps
...
+        IBeatmapInfo BeatmapInfo { get; }
+
+        IBeatmapSetInfo BeatmapSetInfo => BeatmapInfo.BeatmapSet;
+
+        IBeatmapMetadataInfo Metadata => BeatmapInfo.Metadata;
+
+        /// <summary>
+        /// Whether the Beatmap has finished loading.
+        ///</summary>
+        public bool BeatmapLoaded { get; }
```
Comment (smoogipoo) [inline osu.Game/Beatmaps/IWorkingBeatmap.cs:29]: How do these `public`s sometimes appear and disappear? I don't mention them anymore because it's a lost cause, but is this some sort of automatic refactor doing it?

Comment (peppy) [reply osu.Game/Beatmaps/IWorkingBeatmap.cs:29]: It's manual - I copy the methods across from non-interface to interface and forget to remove the `public` prefix.

File: osu.Game/Beatmaps/IWorkingBeatmap.cs [hunk 3]
```diff
@@ -79,5 +115,20 @@ public interface IWorkingBeatmap
         /// </summary>
         /// <param name="storagePath">The storage path to the file.</param>
         Stream GetStream(string storagePath);
+
+        /// <summary>
+        /// Beings loading the contents of this <see cref="IWorkingBeatmap"/> asynchronously.
+        /// </summary>
+        public void BeginAsyncLoad();
```
Comment (smoogipoo) [inline osu.Game/Beatmaps/IWorkingBeatmap.cs:122]: I dunno about this stuff... The whole "load" stuff is very game-specific.

My hope is to eventually revert these changes after you're done with things, so let's go with it as it is.

Comment (peppy) [reply osu.Game/Beatmaps/IWorkingBeatmap.cs:122]: What's the expectation of `IWorkingBeatmap`? No async handling at all? I can get behind that, but needs a lot of refactoring work, at a much later stage.

File: osu.Game/Beatmaps/WorkingBeatmap.cs (hunk 1 above)
Comment (smoogipoo) [reply osu.Game/Beatmaps/WorkingBeatmap.cs:35]: Can you use `new` instead of the disable?

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Beatmaps/WorkingBeatmap.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Beatmaps/WorkingBeatmap.cs:35]: Unfortunately doesn't work for interface methods.

Comment (smoogipoo) [reply osu.Game/Beatmaps/WorkingBeatmap.cs:35]: How do you feel about just implementing explicitly:
```diff
diff --git a/osu.Game/Beatmaps/WorkingBeatmap.cs b/osu.Game/Beatmaps/WorkingBeatmap.cs
index 51eea94d3a..682eaded1e 100644
--- a/osu.Game/Beatmaps/WorkingBeatmap.cs
+++ b/osu.Game/Beatmaps/WorkingBeatmap.cs
@@ -27,11 +27,7 @@ namespace osu.Game.Beatmaps
     public abstract class WorkingBeatmap : IWorkingBeatmap
     {
         public readonly BeatmapInfo BeatmapInfo;
-
-        // ReSharper disable once FieldHidesInterfacePropertyWithDefaultImplementation
         public readonly BeatmapSetInfo BeatmapSetInfo;
-
-        // ReSharper disable once FieldHidesInterfacePropertyWithDefaultImplementation
         public readonly BeatmapMetadata Metadata;
 
         protected AudioManager AudioManager { get; }
@@ -232,6 +228,10 @@ private Task<IBeatmap> loadBeatmapAsync()
 
         IBeatmapInfo IWorkingBeatmap.BeatmapInfo => BeatmapInfo;
 
+        IBeatmapSetInfo IWorkingBeatmap.BeatmapSetInfo => BeatmapSetInfo;
+
+        IBeatmapMetadataInfo IWorkingBeatmap.Metadata => Metadata;
+
         public IBeatmap Beatmap
         {
             get
```
Otherwise, I think there may be a nullability issue here - doesn't `((IWorkingBeatmap)beatmap).Metadata` violate the non-nullable constraint of `IBeatmapInfo.Metadata` given the `WorkingBeatmap` ctor:
```
BeatmapInfo = beatmapInfo;
Metadata = beatmapInfo.Metadata ?? BeatmapSetInfo?.Metadata ?? new BeatmapMetadata();
```

Comment (peppy) [reply osu.Game/Beatmaps/WorkingBeatmap.cs:35]: Yeah, I'm fine with doing that. Will make the changes.

**Review by smoogipoo — APPROVED** [review]
